pkg/cli/adbtunnel.go

```diff
@@ -223,48 +236,256 @@ func (f *ADBForwarder) Send(data []byte) error {
 	return nil
 }
 
-func (f *ADBForwarder) acceptLoop(listener net.Listener) {
-	defer listener.Close()
+type StatusMsg struct {
+	ServiceURL string `json:"service_url"`
+	Zone       string `json:"zone"`
+	Host       string `json:"host"`
+	Device     string `json:"device"`
+	Port       int    `json:"port"`
+	Status     string `json:"status"`
+}
+
+func (f *ADBForwarder) StatusJSON() []byte {
+	f.statusMtx.Lock()
+	status := f.status
+	f.statusMtx.Unlock()
+
+	msg := StatusMsg{
+		ServiceURL: f.serviceURL,
+		Zone:       f.zone,
+		Host:       f.host,
+		Device:     f.device,
+		Port:       f.port,
+		Status:     status,
+	}
+	ret, err := json.Marshal(&msg)
+	if err != nil {
+		panic("Couldn't marshal status map")
+	}
+	return ret
+}
+
+func (f *ADBForwarder) setStatus(status string) {
+	f.statusMtx.Lock()
+	defer f.statusMtx.Unlock()
+	f.status = status
+}
+
+func (f *ADBForwarder) acceptLoop() {
+	f.statusMtx.Lock()
+	if f.status != ADBFwdReady {
+		// This function should always be called from StartForwarding
+		f.logger.Printf("Forwarder accept loop started in wrong state: %s", f.state)
+	}
+	f.statusMtx.Unlock()
+
+	defer f.listener.Close()
 	for {
-		conn, err := listener.Accept()
-		if !f.running.Load() {
-			// StopForwarding was called, Accept likely returned an error because
-			// listener was closed, ignore it.
+		conn, err := f.listener.Accept()
+		if err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.logger.Printf("Error accepting connection on port %d: %v", f.port, err)
+			}
 			return
 		}
+		f.statusMtx.Lock()
 		f.conn = conn
```

Comment:
nit: extract this into a new smaller method: `setConnection`

Reply:
Done
Great suggestion! I also added `compareAndSwapStatus(old, new)` to change states atomically.